lib/spack/docs/environments.rst

```diff
@@ -0,0 +1,322 @@
+.. Copyright 2013-2018 Lawrence Livermore National Security, LLC and other
+   Spack Project Developers. See the top-level COPYRIGHT file for details.
+
+   SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+.. _environments:
+
+Environments
+============
+
+An environment is used to group together a set of specs for the
+purpose of building, rebuilding and deploying in a coherent fashion.
+Environments provide a number of advantages over the a la carte
+approach of building and loading individual Spack modules:
+
+#. Environments eliminate the non-determinism encountered when
+   installing packages and then loading what was just created.
+   Consider the following sequence:
+
+   .. code-block:: console
+
+      $ spack install mypackage
+      $ spack load mypackage
+
+   Sometime later, somebody else installs an earlier version of ``mypackage``:
+
+   .. code-block:: console
+
+      $ spack install mypackage@0.9
+
+   Now the ``spack load mypackage`` command will fail because more
+   than one version of ``mypackage`` is installed.  Thus, ``spack
+   load`` is non-deterministic, and not a reliable way to load
+   specific sets of modules one wishes to build and load.
+   Environments solve this problem by creating a localized context
+   that is immune to what other versions of packages might be
+   installed in the global Spack prefix.
+#. Environments separate the steps of (a) choosing what to
+   install, (b) concretizing, and (c) installing.  This allows
+   Environments to remain stable and repeatable, even if Spack packages
+   are upgraded: specs are only re-concretized when the user
+   explicitly asks for it.  It should even be possible to reliably
+   transport environments between different computers running
+   different versions of Spack!
+#. Since re-concretization only occurs at controlled times, the
+   Environment equivalent of many Spack commands are far faster.  For
+   example, ``spack env <env> list -Ilr`` vs. ``spack spec``.
+#. Environments allow several specs to be built at once; a more robust
+   solution than ad-hoc scripts making multiple calls to ``spack
+   install``.
+#. An Environment that is built as a whole can be loaded as a whole.
+   Spack can generate a script of ``module load`` commands that load
+   the appropriate environment.  And that script will work, without
+   running Spack, as long as the underlying modules remain installed.
```

Comment:
@citibeth: FYI, I really want `spack activate` to load the entire environment *without* modules, so that you can use Spack the same way everywhere without a lot of bootstrapping.  So I basically agree with this and think virtual envs are a better interface than piecemeal modules.